### ISSUE TITLE
Files extracted from zip are currupt

### ISSUE THREAD

**Issue by webwake** [ZJONSSON/node-unzipper#271]
The extracted files are getting corrupts, this started to happen after upgrading from Node 18 to Node.js 19.8.1 on Windows 10 (haven't tested if broader issue)

Below is a code sample to reproduce the issue. Have incorporated the decompress library to demonstrate an non-corrupt file

```
const fs = require('fs');
const unzipper = require('unzipper');
const { https } = require('follow-redirects');
const crypto = require("crypto");

const decompress = require('decompress');

async function downloadBinariesFromRelease() {
    await new Promise((resolve, reject) => {
        fs.mkdirSync('unzipper', { recursive: true });
        const file = fs.createWriteStream('binaries.zip');
        console.log('Downloading Neutralinojs binaries..');
        https.get("https://github.com/neutralinojs/neutralinojs/releases/download/v4.10.0/neutralinojs-v4.10.0.zip", function (response) {
            response.pipe(file);
            response.on('end', () => {
                resolve();
            });
        });
    });
}

function fileHash(filename, algorithm = 'md5') {
    return new Promise((resolve, reject) => {
        // Algorithm depends on availability of OpenSSL on platform
        // Another algorithms: 'sha1', 'md5', 'sha256', 'sha512' ...
        let shasum = crypto.createHash(algorithm);
        try {
            let s = fs.ReadStream(filename)
            s.on('data', function (data) {
                shasum.update(data)
            })
            // making digest
            s.on('end', function () {
                const hash = shasum.digest('hex')
                return resolve(hash);
            })
        } catch (error) {
            return reject('calc fail');
        }
    });
}

async function extractWithUnzipperLibrary() {
    await new Promise((resolve, reject) => {
        console.log('Extracting using "unzipper" library')
        fs.createReadStream('binaries.zip')
            .pipe(unzipper.Extract({ path: './unzipper' }))
            .promise()
            .then(() => resolve())
            .catch((e) => reject(e));
    });
}

async function extractWithDecompressLibrary() {
    console.log('Extracting using "decompress" library')
    await decompress('binaries.zip', 'decompress');
}


async function main() {
    await downloadBinariesFromRelease();
    await extractWithUnzipperLibrary();
    await extractWithDecompressLibrary();
    

    console.log(`correct hash:    3bbb562a59a454534f0ced6c801ccdb7`);
    console.log(`unzipper hash:   ${await fileHash("unzipper/neutralino-win_x64.exe", "md5")}`);
    console.log(`decompress hash: ${await fileHash("decompress/neutralino-win_x64.exe", "md5")}`);
}

main();

```

**Comment by mvolfik** [comment]
I have another code sample. Unzipping works when you send the zip file as one large chunk, but not when you slice it like this (I discovered this by piping the file from https download, so I suppose these are some transmission window sizes).

Code:

```js
const fs = require("node:fs");
const unzipper = require("unzipper");

const data = fs.readFileSync("a.zip");
const extractor = unzipper.Extract({ path: "./test" });

extractor.write(Uint8Array.prototype.slice.call(data, 0, 1378));
extractor.write(Uint8Array.prototype.slice.call(data, 1378, 1378 * 2));
extractor.write(Uint8Array.prototype.slice.call(data, 1378 * 2, 1378 * 3));
extractor.write(Uint8Array.prototype.slice.call(data, 1378 * 3));
```

The zipfile: [a.zip](https://github.com/ZJONSSON/node-unzipper/files/11240627/a.zip)

**Comment by matej-marcisovsky** [comment]
> 

Downgrading to 18.14.0 fixed issues for me.

**Comment by didaquis** [comment]
Hi @mvolfik and @webwake.  

Can either of you confirm what I say in my previous comment? In my case using Node 18.15.0 no error occurs. However, the files are corrupted if Node 18.16.0 is used.

**Comment by didaquis** [comment]
May be related to https://github.com/ZJONSSON/node-unzipper/issues/269

**Comment by Walther** [comment]
Unfortunately, this seems to be reproducible using Node 18.16.0 even with `unzipper` version `0.10.14`, which was released after merging https://github.com/ZJONSSON/node-unzipper/pull/274

**Comment by sounisi5011** [comment]
> Is there an update to fix version 18.16 ?

It probably does not exist. You may want to migrate to another package. A list of candidates I have researched is written here: https://github.com/go-task/go-npm/issues/7#issuecomment-1568567117

**Comment by thorsent** [comment]
The problem appears to be that unzipper is putting the first block of code in the wrong place.

This is a winmerge comparison of a correctly unzipped file (left) and the results of unzipper on latest node (right). It appears that the very first block from the file has been written to disk later than it should. The resulting file has the same file size but is corrupt.
![image](https://github.com/ZJONSSON/node-unzipper/assets/1657084/9618a465-75cc-4c7e-913d-ffd89a4b9ca5)

**Comment by ericman314** [comment]
> The problem appears to be that unzipper is putting the first block of code in the wrong place.

Same thing is happening on Node 20.7.0, using unzipper version 0.10.14. In my case it's usually blocks of 2^14 bytes that are being swapped, but it also happens with files much smaller than this.

Would love to see this fixed, as Node 16 has reached end-of-life. There are alternatives, but I really like unzipper's api best, especially the ability to stream archived data without loading it into a buffer first. Very cool!

**Comment by pixartist** [comment]
Very simple application here, all data corrupted by this library

**Comment by sovcik** [comment]
Same issue here.

**Comment by wfairclough** [comment]
Also same issue, if I didn't find this page I would have wasted many more hours debugging.

**Comment by Lulalaby** [comment]
CR v20.10.0

**Comment by DaCao** [comment]
same issue here. all data corrupted by this library. 
Wasted 4 days on this.... 

please fix it.

**Comment by kenotron** [comment]
https://www.npmjs.com/package/yauzl#no-streaming-unzip-api - wondering if that's the reason?

**Comment by samerkassem82** [comment]
So is this library dead? Same issas as almost a year later

**Comment by dy-dx** [comment]
The issue has been fixed in the following node.js versions:
- [v18.20.0](https://nodejs.org/en/blog/release/v18.20.0)
- [v20.12.0](https://nodejs.org/en/blog/release/v20.12.0)
- [v21.7.1](https://nodejs.org/en/blog/release/v21.7.1)

But for anyone unable to upgrade, you'll need to switch to a different unzipper library until this gets addressed: https://github.com/ZJONSSON/node-unzipper/issues/261

**Comment by ZJONSSON** [comment]
Thanks @dy-dx for identifying the underlying issue as a bug with `fs.createWriteStream` in nodejs.  This has now been fixed in node per https://github.com/ZJONSSON/node-unzipper/issues/271#issuecomment-2021223739

As far as extract goes we have moved from unmaintained `fstream` to `fs-extra` in a newly published version (`v0.12.1`)
